server/src/main/java/io/crate/replication/logical/LogicalReplicationService.java

```diff
@@ -301,44 +344,125 @@ public void verifyTablesDoNotExist(String subscriptionName,
                 Strings.EMPTY_ARRAY
             );
 
-        FutureActionListener<RestoreService.RestoreCompletionResponse, RestoreService.RestoreCompletionResponse> future =
+        var finalFuture = new CompletableFuture<RestoreService.RestoreCompletionResponse>();
+        FutureActionListener<RestoreService.RestoreCompletionResponse, RestoreService.RestoreCompletionResponse> restoreFuture =
             FutureActionListener.newInstance();
-        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new AbstractRunnable() {
-            @Override
-            public void onFailure(Exception e) {
-                future.onFailure(e);
+        restoreFuture.whenComplete(
+            (response, err) -> {
+                if (err == null) {
+                    updateSubscriptionState(
+                        subscriptionName,
+                        Subscription.State.RESTORING,
+                        null
+                    ).thenAccept(ignored -> finalFuture.complete(response));
+                } else {
+                    updateSubscriptionState(
+                        subscriptionName,
+                        Subscription.State.FAILED,
+                        "Failed restoring subscription, error=" + err.getMessage()
+                    ).whenComplete((ignored, ignoredErr) -> finalFuture.completeExceptionally(err));
+                }
             }
+        );
```

Comment:
Would it be possible to use the result of `whenComplete` as `finalFuture` and remove the inner `.thenAccept` / `.whenComplete` ?

Took me a moment to realize that `finalFuture` is chained to the `resultFuture`

Reply:
Maybe I miss something but when using the result of the `whenComplete`, the values set on the future are the results of the `updateSubscriptionState` and not the original ones of the `restoreFuture` which we are interested in.